tracer/src/Datadog.Trace/Propagators/W3CTraceContextPropagator.cs

```diff
@@ -90,25 +89,30 @@ internal static string CreateTraceParentHeader(SpanContext context)
 
         internal static string CreateTraceStateHeader(SpanContext context)
         {
-            var samplingPriority = SamplingPriorityToString(context.TraceContext?.SamplingPriority ?? context.SamplingPriority);
-            StringBuilder? sb = null;
+            var sb = StringBuilderCache.Acquire(100);
 
             try
             {
-                sb = StringBuilderCache.Acquire(100);
                 sb.Append("dd=");
 
+                // sampling priority ("s:<value>")
+                var samplingPriority = SamplingPriorityToString(context.TraceContext?.SamplingPriority);
+
                 if (samplingPriority != null)
                 {
                     sb.Append("s:").Append(samplingPriority).Append(';');
                 }
 
-                if (!string.IsNullOrWhiteSpace(context.Origin))
+                // origin ("o:<value>")
+                var origin = context.TraceContext?.Origin;
```

Comment:
The real source of truth for these trace values (sampling priority, origin, propagated tags, and now the additional W3C trace state) is the `TraceContext`. A few properties in `SpanContext` have that fallback for backwards compat reasons because there is old code that still reads or sets this values in `SpanContext`.

For example, until recently there was no `TraceContext.Origin`, so we kept copying origin from one `SpanContext` to the next, for every span in a trace. Now `SpanContext.Origin` should only have a value when extracting from headers, and we copy that to `TraceContext.Origin` and use only that for the rest of the trace. This is part of the `SpanContext` cleanup I keep saying I want to do to remove all this source of confusion.

Reply:
To my point: this change just uncovered an issue with some tests and made them start failing. Those tests were relying on the wrong use of `SpanContext` and not testing the code paths we actually used, so I am now fixing those tests.